utilities/cli/util.py

```diff
@@ -319,6 +391,61 @@ def __init__(self, package_name: str, version_pattern: str, message: str = None)
         )
 
 
+def is_package_installed(package_name: str) -> bool:
+    """Check if a package is already installed"""
+    try:
+        result = subprocess.run(
+            ["dpkg", "-l", package_name], capture_output=True, text=True, check=False
+        )
+        return result.returncode == 0 and "ii" in result.stdout
+    except (subprocess.CalledProcessError, FileNotFoundError):
+        return False
+
+
+def ensure_apt_updated(dry_run: bool = False) -> None:
+    """Ensure apt package list is updated, but only once per session"""
```

Comment:
The risk of missing some update in some long-running session might outweigh the benefit of enforcing "only once" here.

Could we run `apt update` once as the first command in sequence, and document an implicit assumption in subsequent package installs that the apt cache must already be updated? Note that `apt install` already makes the same assumption.